### PR TITLE
mips32: clobber memory on sycall

### PR DESCRIPTION
**Pull request by vhotspur**
As the syscall may touch arbitrary memory, we need to prevent the compiler to cache some values in registers.

Actual problem was demonstrated on Kalisto during school assignments (thanks Adam Frey) but HelenOS uses basically the same code, thus the issue applies here too.

Can someone better versed with other architectures check that this does not apply to them as well?

### CODE REVIEW

**Comment by le-jzr** [comment]
I think all the registers that are not explicitly preserved by the kernel must be marked clobbered/modified as well.

**Comment by le-jzr** [comment]
Does it? Hmm. In general the kernel side of syscall doesn't need to preserve scratch registers. I think we might be doing it for debugging reasons, but still it would be safer (more future-proof) if the userspace portion works even if we decide not to store them and instead just clear them on return (as an optimization).

**Comment by le-jzr** [comment]
Though there should be no expectation that the syscall follows standard calling convention, so that's not really relevant. Ignore me.

File: uspace/lib/c/arch/mips32/src/syscall.c [hunk 1]
```diff
@@ -60,7 +60,12 @@ sysarg_t __syscall(const sysarg_t p1, const sysarg_t p2, const sysarg_t p3,
 	       * We are a function call, although C
 	       * does not know it.
 	       */
-	    : "%ra"
+	    : "%ra",
```
Comment (jermar) [inline uspace/lib/c/arch/mips32/src/syscall.c]: To make things consistent, the colon should now (that there are two clobbers) go before the comment.

**Comment by jermar** [comment]
IMHO arm32 and ppc32 need the same fix.

**Comment by jermar** [comment]
An interesting case is what happens on architectures that have the syscall wrapper implemented in a .S file in an extern function. From that I'd assume that calling a function represents an implicit memory clobber. It certainly does represent a 'sequence point' from the point of the language

**Comment by le-jzr** [comment]
Yes, external function that's not declared with `__attribute__((pure))` is always assumed to clobber memory.

**Comment by jxsvoboda** [comment]
Am I right thinking that this can only cause problems with LTO? Because otherwise any incorrect optimizations are contained to the syscall function. When the function is used, without LTO, it is an external function and the compiler will assume it clobbers memory. With LTO the linker will have the information that it is not supposed to clobber memory.

**Comment by le-jzr** [comment]
It doesn't follow. Functions are assumed to clobber memory because they are opaque to the caller. When LTO is involved, the compiler can consider internal details of the function. Inline assembler needs the clobber information explicitly, because gcc can't understand the assembly. In principle, this can affect even assembly generated by gcc in prologue/epilogue of otherwise empty function. In practice it doesn't seem to, but it can.

**Comment by vhotspur** [comment]
> IMHO arm32 and ppc32 need the same fix.

Added.

Any objections for merging this?

**Pull request merged**